### PR TITLE
NETOBSERV-395: Tls loki

### PR DESCRIPTION
**Pull request by OlivierCazade**
Add tls configuration to loki using normal TLS (not mTLS) and also supporting a flag to skip TLS checks.

### CODE REVIEW

File: controllers/consoleplugin/consoleplugin_objects.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ const configMapName = "console-plugin-config"
 const configFile = "config.yaml"
 const configVolume = "config-volume"
 const configPath = "/opt/app-root/"
+const lokiCerts = "kafka-certs"
```
Comment (jpinsonneau) [inline controllers/consoleplugin/consoleplugin_objects.go:28]:
```suggestion
const lokiCerts = "loki-certs"
```

File: controllers/consoleplugin/consoleplugin_objects.go [hunk 2]
```diff
@@ -112,6 +113,55 @@ func buildArgs(desired *flowsv1alpha1.FlowCollectorConsolePlugin, desiredLoki *f
...
+			ReadOnly:  true,
+		}}
+
+	args := []string{
+		"-cert", "/var/serving-cert/tls.crt",
+		"-key", "/var/serving-cert/tls.key",
+		"-loki", querierURL(b.desiredLoki),
+		"-loki-labels", strings.Join(constants.LokiIndexFields, ","),
+		"-loki-tenant-id", b.desiredLoki.TenantID,
+		"-loki-skip-tls", "true",
```
Comment (jpinsonneau) [inline controllers/consoleplugin/consoleplugin_objects.go:152]:
```suggestion
```

Comment (jpinsonneau) [reply controllers/consoleplugin/consoleplugin_objects.go:152]: Do you also experience [this issue](https://issues.redhat.com/browse/NETOBSERV-485) on args order ?

Comment (OlivierCazade) [reply controllers/consoleplugin/consoleplugin_objects.go:152]: Either I was lucky and it kept the same order, or I did not have the issue.

Using this configuration I was able to reach a loki with tls:
`    tls:
      enable: true
      insecureSkipVerify: true
`

File: config/samples/flows_v1alpha1_flowcollector.yaml [hunk 3]
```diff
@@ -47,14 +47,20 @@ spec:
         certFile: user.crt
         certKey: user.key
   loki:
-    url: 'http://loki:3100/'
+    url: 'http://loki.network-observability.svc:3100/'
```
Comment (jotak) [inline config/samples/flows_v1alpha1_flowcollector.yaml:50]: did you need this change / does it makes anything easier?
The problem I see by inferring namespace is that if you change the `spec.namespace`, then this url is likely invalid.
By sticking with just `http://loki` we assume loki is deployed next to the other components

File: controllers/consoleplugin/consoleplugin_objects.go [hunk 4]
```diff
@@ -112,6 +113,55 @@ func buildArgs(desired *flowsv1alpha1.FlowCollectorConsolePlugin, desiredLoki *f
...
+		"-loglevel", b.desired.LogLevel,
+		"-frontend-config", configPath + configFile}
+
+	if b.desiredLoki != nil && b.desiredLoki.TLS.Enable {
```
Comment (jotak) [inline controllers/consoleplugin/consoleplugin_objects.go:156]: Maybe I'm wrong but I think the `InsecureSkipVerify` and the CA cert shouldn't be mutually exclusive. The loki server might expect TLS encryption (thus needs CA cert), but still, `InsecureSkipVerify` can be used (in this case it is the client that skips certificate verification, but that doesn't mean it wouldn't encrypt communications).
At least that's my understanding. So if that's correct, there shouldn't be an `else` clause here for CA.

File: controllers/flowlogspipeline/flp_objects.go [hunk 5]
```diff
@@ -338,11 +343,20 @@ func (b *builder) addTransformStages(lastStage *config.PipelineBuilderStage) {
...
-				InsecureSkipVerify: true,
-			},
+		if b.desiredLoki != nil && b.desiredLoki.TLS.Enable {
+			if b.desiredLoki.TLS.InsecureSkipVerify {
+				lokiWrite.ClientConfig = &promConfig.HTTPClientConfig{
+					TLSConfig: promConfig.TLSConfig{
+						InsecureSkipVerify: true,
+					},
+				}
+			} else {
```
Comment (jotak) [inline controllers/flowlogspipeline/flp_objects.go:353]: Same remark here, I don't think there should be a `else`

File: config/samples/flows_v1alpha1_flowcollector.yaml (hunk 3 above)
Comment (OlivierCazade) [reply config/samples/flows_v1alpha1_flowcollector.yaml:50]: Yes, the certificate generated by Openshift is only valid for the full name.

File: controllers/consoleplugin/consoleplugin_objects.go (hunk 4 above)
Comment (OlivierCazade) [reply controllers/consoleplugin/consoleplugin_objects.go:156]: InsecureSkipVerify does not mean not using TLS it means no checking that the certificate used by the remote server is properly signed.
The CA cert is used for the same reason, it is used to validate the remote server certificate.

File: config/samples/flows_v1alpha1_flowcollector.yaml (hunk 3 above)
Comment (jotak) [reply config/samples/flows_v1alpha1_flowcollector.yaml:50]: Ah I see, ok then

File: controllers/consoleplugin/consoleplugin_objects.go (hunk 4 above)
Comment (jotak) [reply controllers/consoleplugin/consoleplugin_objects.go:156]: > InsecureSkipVerify does not mean not using TLS it means no checking that the certificate used by the remote server is properly signed.

That's what I say :)

>  The CA cert is used for the same reason, it is used to validate the remote server certificate. 

But the CA cert contains the server public key. So if TLS is enabled then I think you _need_ this cert for encryption, even if InsecureSkipVerify is true

Comment (jotak) [reply controllers/consoleplugin/consoleplugin_objects.go:156]: (If this is wrong then I need to fix that on Kafka side, because that's how I did)

Comment (jotak) [reply controllers/consoleplugin/consoleplugin_objects.go:156]: Ok, I've read a little more about TLS and I think you're right :)

**Review by jotak — APPROVED** [review]

**Pull request merged**